### ISSUE TITLE
meta issue - faster/more efficient fast sync

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#952]
Tracking a variety of things here around faster "fast sync".
Some of this is probably premature in terms of optimization but tracking here as some may make sense earlier.

At height ~24,000 we are asking a peer for a txhashset.zip that is approx 30MB in size.
(Although this is based purely on how pruned the MMRs are on the sending peer).
Downloading the zip over a slowish connection takes a few minutes.
Verifying all those rangeproofs takes several minutes.

I'd like to propose the following as possible enhancements to this -

* ask for the 3 MMRs separately (smaller zip files, possibly parallelizable?)
  * outputs
  * kernels
  * rangeproofs
* send over the following per MMR (tracking this in issue [tbd]) -
  * data file (pruned and compacted)
  * pruned/compacted hashes (reconstruct hash file on receipt)
* potentially process rangeproofs separately (likely a premature optimization right now)?
* merge latest secp256k1 lib and api
* leverage batch rangeproof verification

Questions - 
* Can we download the txhashset faster if we get it in smaller parts from multiple peers?
  * Is it safe to split this up and request from multiple peers (if different peers have potentually pruned at different pos)?
* Can we start asking for blocks earlier? What data do we need locally before we can start asking for blocks at or near the horizon?
* Can we work with a shorter horizon (to minimize blocks required) and incrementally add more blocks?
  * i.e. Can we get the node to a running state sooner (possibly running the risk of failing to handle a fork, even if temporary?)
* Can we take a "fast sync non-archival node" and incrementally turn it back into a full archival node?
  * i.e. Do we make fast sync not just the default but the _only_ way of syncing? Would we want to do this?
  * run fast sync, then request all the blocks, replacing txhashset with our own data?
* Can we get chunks of an MMR (first 1,000 leaves, next 1,000 leaves etc.)?
  * This may allow us to get an MMR incrementally and we could verify parts of it while we are downloading subsequent chunks

**Comment by DavidBurkett** [comment]
Also, BZip2 is one of the slower compression algorithms, and for sufficiently random data like we have (hashes, for example), it doesn't actually noticeably decrease the size of the TxHashSet. A TxHashSet I just downloaded compressed 45,758,419B to 44,989,282B, a savings of <2%. In fact, some of the bin files inside actually took 10-20% more space when 'compressed'. It'd probably be better to just send the raw data, and replace general compression with something more effective like sending leaves-only.

**Comment by hashmap** [comment]
Any objections to close it? We could have some more specific issues if needed.

**Comment by antiochp** [comment]
Yeah let's close this and keep our issue list clean. "Minimizing inventory".
